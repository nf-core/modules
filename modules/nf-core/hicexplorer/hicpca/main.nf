process HICEXPLORER_HICPCA {
    tag "$meta.id"
    label 'process_medium'

    conda "bioconda::hicexplorer=3.7.2 numpy=1.23.5"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/hicexplorer:3.7.2--pyhdfd78af_1':
        'biocontainers/hicexplorer:3.7.2--pyhdfd78af_1' }"

    input:
    tuple val(meta), path(matrix)

    output:
    tuple val(meta), path("${prefix}_*")           , emit:results
    tuple val(meta), path("${prefix}_pca1.$format"), emit:pca1
    tuple val(meta), path("${prefix}_pca2.$format"), emit:pca2
    path("versions.yml")                           , emit:versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    prefix   = task.ext.prefix ?: "${meta.id}"
    args     = args.tokenize()
    def idx  = args.findIndexOf{ it == '--format' | it == '-f' }
    format   = 'bigwig'
    if (idx>=0) {
        format = args[idx+1]
        args.remove(idx+1)
        args.remove(idx)
    }
    idx = args.indexOf('--whichEigenvectors')
    eigenvectors = '1 2'
    if(idx>=0) {
        eigenvectors = args[idx + 1]
        args.remove(idx+1)
        args.remove(idx)
    }
    outfilenames = eigenvectors.tokenize()
        .collect{"${prefix}_pca${it}.${format}"}.join(' ')
    args = args.join(' ')
    """
    hicPCA \\
        -m $matrix \\
        $args \\
        --format $format \\
        --whichEigenvectors $eigenvectors \\
        --outputFileName $outfilenames

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        hicexplorer: \$(hicPCA --version 2>&1 | sed 's/hicPCA //')
    END_VERSIONS
    """
}
