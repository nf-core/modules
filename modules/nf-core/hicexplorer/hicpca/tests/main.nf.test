nextflow_process {

    name "Test Process HICEXPLORER_HICPCA"
    script "../main.nf"
    process "HICEXPLORER_HICPCA"

    tag "modules"
    tag "modules_nfcore"
    tag "hicexplorer"
    tag "hicexplorer/hicpca"

    test("mm9 chr1") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test' ], // meta map
                    file('https://github.com/deeptools/HiCExplorer/raw/master/hicexplorer/test/test_data/hicPCA/mm9_reduced_chr1.cool', checkIfExists: true)
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("mm9 chr1 - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [ id:'test' ], // meta map
                    file('https://github.com/deeptools/HiCExplorer/raw/master/hicexplorer/test/test_data/hicPCA/mm9_reduced_chr1.cool', checkIfExists: true)
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
