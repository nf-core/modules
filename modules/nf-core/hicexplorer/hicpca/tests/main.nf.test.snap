{
    "mm9 chr1": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_pca1.bigwig:md5,e228fde481bedea641bcd0204bb663a4",
                            "test_pca2.bigwig:md5,109c0d3cc70ecaf3647f3293504a9683"
                        ]
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test_pca1.bigwig:md5,e228fde481bedea641bcd0204bb663a4"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test_pca2.bigwig:md5,109c0d3cc70ecaf3647f3293504a9683"
                    ]
                ],
                "3": [
                    "versions.yml:md5,54775e8cd6ce193be8705000c3279192"
                ],
                "pca1": [
                    [
                        {
                            "id": "test"
                        },
                        "test_pca1.bigwig:md5,e228fde481bedea641bcd0204bb663a4"
                    ]
                ],
                "pca2": [
                    [
                        {
                            "id": "test"
                        },
                        "test_pca2.bigwig:md5,109c0d3cc70ecaf3647f3293504a9683"
                    ]
                ],
                "results": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_pca1.bigwig:md5,e228fde481bedea641bcd0204bb663a4",
                            "test_pca2.bigwig:md5,109c0d3cc70ecaf3647f3293504a9683"
                        ]
                    ]
                ],
                "versions": [
                    "versions.yml:md5,54775e8cd6ce193be8705000c3279192"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.3"
        },
        "timestamp": "2025-01-09T11:04:14.308469361"
    },
    "mm9 chr1 - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_pca1.bigwig:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_pca2.bigwig:md5,d41d8cd98f00b204e9800998ecf8427e"
                        ]
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test_pca1.bigwig:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test_pca2.bigwig:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    "versions.yml:md5,54775e8cd6ce193be8705000c3279192"
                ],
                "pca1": [
                    [
                        {
                            "id": "test"
                        },
                        "test_pca1.bigwig:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "pca2": [
                    [
                        {
                            "id": "test"
                        },
                        "test_pca2.bigwig:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "results": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_pca1.bigwig:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_pca2.bigwig:md5,d41d8cd98f00b204e9800998ecf8427e"
                        ]
                    ]
                ],
                "versions": [
                    "versions.yml:md5,54775e8cd6ce193be8705000c3279192"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.3"
        },
        "timestamp": "2025-01-09T11:04:26.856478191"
    }
}